matchzoo/layers/match_layer.py

```diff
@@ -0,0 +1,129 @@
+"""An implementation of Match Layer."""
+from __future__ import absolute_import
+
+import typing
+from keras import backend as K
+from keras.engine import Layer
+
+
+class MatchLayer(Layer):
+    """Layer that computes a matching matrix between samples in two tensors."""
+
+    def __init__(self, normalize: bool = False,
+                 match_type: str = 'dot', **kwargs):
+        """
+        :class:`MatchLayer` constructor.
+
+        :param normalize: Whether to L2-normalize samples along the
+            dot product axis before taking the dot product.
+            If set to True, then the output of the dot product
+            is the cosine proximity between the two samples.
+        :param **kwargs: Standard layer keyword arguments.
+        """
+        super(MatchLayer, self).__init__(**kwargs)
+        self.normalize = normalize
+        self.match_type = match_type
+        self.supports_masking = True
+        if match_type not in ['dot', 'mul', 'plus', 'minus', 'concat']:
+            raise ValueError('In `MatchLayer` layer, '
+                             'match_type={} is unknown.'.format(match_type))
+
+    def build(self, input_shape: list):
+        """
+        Build the layer.
+
+        :param input_shape: the shapes of the input tensors, for MatchLayer we
+            need tow input tensors.
+        """
+        # Used purely for shape validation.
+        if not isinstance(input_shape, list) or len(input_shape) != 2:
+            raise ValueError('A `MatchLayer` layer should be called '
+                             'on a list of 2 inputs.')
+        self.shape1 = input_shape[0]
+        self.shape2 = input_shape[1]
+        if self.shape1[0] != self.shape2[0]:
+            raise ValueError(
+                'Dimension incompatibility '
```

Comment:
dimensional incompatibility or incompatible dimensions, personally think the later is better.